src/lib.rs

```diff
@@ -162,3 +219,206 @@ impl LevelIndex {
         }
     }
 }
+
+bitflags::bitflags! {
+    #[repr(transparent)]
+    pub struct ChannelTypeQualifiers: u32 {
+        const LINEAR        = (1 << 0);
+        const EXPONENT      = (1 << 1);
+        const SIGNED        = (1 << 2);
+        const FLOAT         = (1 << 3);
+    }
+}
+
+bitflags::bitflags! {
+    #[derive(Default)]
+    #[repr(transparent)]
+    pub struct DataFormatFlags: u32 {
+        const STRAIGHT_ALPHA             = 0;
+        const ALPHA_PREMULTIPLIED        = (1 << 0);
+    }
+}
+
+// This can be obtained from std::mem::transmute::<f32, u32>(1.0f32) but we
+// want to support nostd. It is used for identifying normalized sample types
+// as in Unorm or Snorm
+const F32_1_AS_U32: u32 = 1065353216;
+
+#[derive(Debug, Default)]
+pub struct BasicDataFormatDescriptor<'data> {
+    pub vendor_id: u32,             //: 17;
+    pub descriptor_type: u32,       //: 15;
+    pub version_number: u32,        //: 16;
```

Comment:
Is it possible for `version_number` to vary without changing the semantics/presence of other fields? If not, we might as well omit it.

Reply:
The Data Format Descriptor blocks always have the same first 8 bytes containing the vendor id, descriptor type, version number, and descriptor block size: https://www.khronos.org/registry/DataFormat/specs/1.3/dataformat.1.3.html#_anchor_id_descriptorblock_xreflabel_descriptorblock_descriptor_block . These are to allow for extension with many different types of data format descriptor to express whatever information someone may need.

Khronos primarily define the Basic Data Format Descriptor that I have implemented parsing for which covers all common compressed and uncompressed texture formats and honestly already much more: https://www.khronos.org/registry/DataFormat/specs/1.3/dataformat.1.3.html#_anchor_id_basicdescriptor_xreflabel_basicdescriptor_khronos_basic_data_format_descriptor_block

Khronos also define two 'advanced' data format descriptors that are extensions to the supported number of dimensions and their corresponding byte planes. For example, the format may have 'planes' due to subsampling of channel data. Y'CbCr 4:2:0 is given as an example where it has dimensions 2x2x1x1 (the basic data format descriptor covers four dimensional textures by default) and the Y' samples are split into two planes which are offset by one line in the y dimension. The data format descriptor for additional byte planes: https://www.khronos.org/registry/DataFormat/specs/1.3/dataformat.1.3.html#_additional_planes_descriptor_block and for additional dimensions: https://www.khronos.org/registry/DataFormat/specs/1.3/dataformat.1.3.html#_additional_dimensions_descriptor_block . I didn't implement either of those as they are out of scope for my needs (2D/3D uncompressed and compressed textures for real-time graphics usage.)

I suppose one could extract out the first 8 bytes of the BasicFormatDataDescriptor struct as a DataFormatDescriptorHeader and always parse it, then you conceptually want to have different types of the main body of the data format descriptor blocks based on what is in the header. The user doesn't need to see the header (necessarily? unless they want to be able to parse the data themselves because it's something that this crate doesn't know how to parse...) Hmm. Maybe the iterator should return something that can either be a BasicDataFormatDescriptor body, or something 'unknown' that contains the DataFormatDescriptorHeader and the body data to support external parsing? Rather than just erroring because this crate doesn't know what the data format descriptor is. I'd want to do it as an enum but I don't know if that is feasible as I think enum members need to be Sized and we won't know what the size of the custom data format descriptor blocks would be until they are parsed.

Reply:
Upon trying it, I don't like the enum approach as different information is needed from each variant and it feels weird to lump it all together. As such, I chose instead to try this:
```rust
pub struct DataFormatDescriptorHeader {
    pub vendor_id: u32,             //: 17;
    pub descriptor_type: u32,       //: 15;
    pub version_number: u32,        //: 16;
    pub descriptor_block_size: u32, //: 16;
}

pub struct DataFormatDescriptor<'data> {
    pub header: DataFormatDescriptorHeader,
    pub data: &'data [u8],
}

pub enum DataFormatDescriptorType {
    Basic, // vendor_id: 0, descriptor_type: 0, version_number: 2
    Unknown,
}

pub struct BasicDataFormatDescriptor<'data> {
    /// None means Unspecified or is an otherwise unknown value
    pub color_model: Option<ColorModel>, //: 8;
    /// None means Unspecified or is an otherwise unknown value
    pub color_primaries: Option<ColorPrimaries>, //: 8;
    /// None means Unspecified or is an otherwise unknown value
    pub transfer_function: Option<TransferFunction>, //: 8;
    pub flags: DataFormatFlags,     //: 8;
    pub texel_block_dimensions: [u32; 4], //: 8 x 4;
    pub bytes_planes: [u32; 8],     //: 8 x 8;
    sample_data: &'data [u8],
}
```

If there is not enough data to parse the header, I return None from the data format descriptors iterator. Otherwise I parse the header. Then if there is not enough data for the block then I return None. Otherwise I create a DataFormatDescriptor from the header and the byte range of the data. The DataFormatDescriptor implements a data_format_descriptor_type function that returns a DataFormatDescriptorType. I would kind of prefer to have a `try_into()` but yet again, not having std prevents that unless I define my own... which I guess I could do. I'm thinking that people can just try_into::<BasicDataFormatDescriptor>() and handle the error. What do you think of this?

Reply:
Alternatively I am back again to wanting to have constructors that return `Option<Self>` or `Result<Self, Something>`

Reply:
Decoupling out the header, which is guaranteed to have a particular structure, and the `BasicDataFormatDescriptor`, which is not necessarily applicable, makes sense to me. Further, is the `descriptor_block_size` useful to the user, or is it redundant to the already suitably sliced `data` field of `DataFormatDescriptor`? Particularly if that field is omitted, we can then derive `Eq + PartialEq` for `DataFormatDescriptorHeader` and define an associated constant corresponding to the header that identifies a `BasicDataFormatDescriptor` which can be easily `match`ed against downstream, and gracefully extended with future descriptor types. This obviates the need for a `data_format_descriptor_type` method and the associated enum without compromising ergonomics.

We could further define pseudo-enums for vendor ID and descriptor type, but because there do not appear to exist any vendor-defined DFDs this is probably premature.

Given that there are a variety of possible present and future descriptor types, not all of which are necessarily known to this library, I think a fallible parsing constructor for `BasicDataFormatDescriptor` is justified. I suggest:

```rust
impl BasicDataFormatDescriptor<'data> {
    pub fn parse(data: &'data [u8]) -> Result<Self, MalformedDfd> { ... }
}

#[derive(Debug, Copy, Clone)]
struct MalformedDfd;
// typical error impls...
```

This does present users with the possibility of a successfully parsed file with malformed DFDs, which is unfortunate; however, because not all DFD types can be known in advance, this is de facto possible regardless as a file with unknown DFDs might be undetectably malformed, so we might as well embrace it. Perhaps @cwfitzgerald would like to weigh in here as well.

Reply:
`descriptor_block_size` is indeed redundant so I removed it, derived `PartialEq + Eq`, defined a constant for the header values for `BasicDataFormatDescriptor`, got rid of `data_format_descriptor_type()` and associated enum.

I definitely like using `parse() -> Result` as a fallible constructor and that leads me to think that the `bytes_to_u32(bytes: &[u8], offset: &mut usize) -> u32` helper that reads 4 bytes into a u32 should itself be fallible and return `Result<u32, ParseError>`, basically returning `ParseError::UnexpectedEnd` if it fails to read the 4 bytes.

Let me know what you think now.

Reply:
I'm a fan of that pattern, since it makes it very hard to forget a bounds check.